algosdk/abi.py

```diff
@@ -0,0 +1,570 @@
+from enum import IntEnum
+import re
+from unittest.mock import Base
+from . import error
+
+
+class BaseType(IntEnum):
+    Uint = 0
+    Byte = 1
+    Ufixed = 2
+    Bool = 3
+    ArrayStatic = 4
+    Address = 5
+    ArrayDynamic = 6
+    String = 7
+    Tuple = 8
+
+
+class Type:
+    """
+    Represents an ABI Type for encoding.
+
+    Args:
+        type_id (BaseType): type of ABI argument, as defined by the BaseType class above.
+        child_type (Type, optional): the type of the child_types array.
+        child_types (list, optional): list of types of the children for a tuple.
+        bit_size (int, optional): size of a uint/ufixed type, e.g. for a uint8, the bit_size is 8.
+        precision (int, optional): number of precision for a ufixed type.
+        static_length (int, optional): index of the asset
+
+    Attributes:
+        type_id (BaseType)
+        child_type (Type)
+        child_types (list)
+        bit_size (int)
+        precision (int)
+        static_length (int)
+    """
+
+    def __init__(
+        self,
+        type_id,
+        child_type=None,
+        child_types=list(),
+        bit_size=None,
+        precision=None,
+        static_length=None,
+    ) -> None:
+        self.abi_type_id = type_id
+        self.child_type = child_type  # Used for arrays
+        self.child_types = child_types  # Used for tuples
+        self.bit_size = bit_size
+        self.precision = precision
+        self.static_length = static_length
+
+    def __str__(self):
+        if self.abi_type_id == BaseType.Uint:
+            return f"uint{str(self.bit_size)}"
```

Comment:
Unfortunately this SDK currently targets Python 3.5+, so f-strings may not be available. Perhaps we should drop support for 3.5 at this point?